docs: testing environment

mdjastrzebski opened the pull request:
### Summary

This docs describes in simple terms how our testing environment operates, so that our users can learn more on behaviour and limitations of using RNTL.

### Test plan

None

AugustinLF commented on website/docs/TestingEnvironment.md:
```diff
@@ -0,0 +1,160 @@
+
+That might sound a bit confusing at first, since we put React Native’s `View` in both categories. There are actually two `View` components: composite one and host one. The relation between them is as follows:
+
+- composite `View` is the type imported from `react-native` package. It’s a JavaScript component, which renderers host `View` as its only child in the element tree.
```
```suggestion
- composite `View` is the type imported from `react-native` package. It’s a JavaScript component, which renders host `View` as its only child in the element tree.
```

AugustinLF commented on website/docs/TestingEnvironment.md:
```diff
@@ -0,0 +1,160 @@
+Any easy way to differentiate between host and composite elements is the `type` prop of `ReactTestInstance`:
+
+- for host components it’s always a string value representing component name, e.g. `"View"`
+- for composite components it’s function or class defining the component
```
```suggestion
- for composite components it’s a function or class corresponding to the component
```

AugustinLF commented on website/docs/TestingEnvironment.md:
```diff
@@ -0,0 +1,160 @@
+
+In the above example user defined components accepts both `onPress` and `style` props but does not pass it (through `Pressable`) to host views, so these props will not affect the user interface.
+
+### Tree navigation
```
Should we add a disclaimer? People should just not use that if they're not 100% they need to. And by that I don't just mean "be careful it's hard", but rather, "you're probably doing something you shouldn't be doing at all"

pierrezimmermannbam replied:
I totally agree, I find it really interesting to explain more what's going on under the hood but the purpose to me is to help people to better use RNTL and to debug more efficiently. I'm not sure about recommending navigating the tree to access host elements, I think good practices should be enforced by the library instead

AugustinLF commented on website/docs/TestingEnvironment.md:
```diff
@@ -0,0 +1,160 @@
+
+### Tree navigation
+
+When navigating three using `parent` or `children` props of `ReactTestInstance` element, you will encounter both host and composite elements. You should be careful when navigating the element tree, as the tree structure for 3rd party components and change independently from your code and cause unexpected test failures.
```
```suggestion
When navigating a tree of react elements using `parent` or `children` props of a `ReactTestInstance` element, you will encounter both host and composite elements. You should be careful when navigating the element tree, as the tree structure for 3rd party components and change independently from your code and cause unexpected test failures.
```

pierrezimmermannbam replied:
I find it really interesting to explain more what's going on under the hood but the purpose to me is to help people to better use RNTL and to debug more efficiently. I'm not sure about recommending navigating the tree to access host elements, I think good practices should be enforced by the library instead

AugustinLF commented on website/docs/TestingEnvironment.md:
```diff
@@ -0,0 +1,160 @@
+
+When you run your tests in React Native Testing Library, somewhat contrary to what the name suggest, they are actually **not** using React Native render. This is because this renderer needs to be run on iOS or Android operating system, so it would need to run on device or simulator.
+
+### React Test Render
```
```suggestion
### React Test Renderer
```

AugustinLF commented on website/docs/TestingEnvironment.md:
```diff
@@ -0,0 +1,160 @@
+
+### React Test Render
+
+Instead, RNTL uses React Test Renderer which is a specialised renderer that allows rending to pure JavaScript objects without access to mobile OS, and can run in Node.js environment using Jest (or other JavaScript test runners). 
```
```suggestion
Instead, RNTL uses React Test Renderer which is a specialised renderer that allows rendering to pure JavaScript objects without access to mobile OS, and that can run in a Node.js environment using Jest (or any other JavaScript test runner). 
```

AugustinLF commented on website/docs/TestingEnvironment.md:
```diff
@@ -0,0 +1,160 @@
+
+Instead, RNTL uses React Test Renderer which is a specialised renderer that allows rending to pure JavaScript objects without access to mobile OS, and can run in Node.js environment using Jest (or other JavaScript test runners). 
+
+Using React Test Renders has both pros and cons.
```
```suggestion
Using React Test Renderer has pros and cons.
```

AugustinLF commented on website/docs/TestingEnvironment.md:
```diff
@@ -0,0 +1,160 @@
+
+Benefits:
+
+- tests can run on most CIs (linux, etc) and do not require mobile device or emulator
```
```suggestion
- tests can run on most CIs (linux, etc) and do not require a mobile device or emulator
```

AugustinLF reviewed (changes requested):
Very clear and well written, nice!

pierrezimmermannbam commented on website/docs/TestingEnvironment.md:
```diff
@@ -0,0 +1,160 @@
+
+## Testing Environment
+
+React Native Testing Library allows you to write integration and component tests for your React Native app or library. While the JSX code used in tests closely resembles your React Native app, the things are not quite simple as they might appear. In this document we will describe the key elements of our testing environment and highlight things to be aware when writing more advanced tests or diagn ...
```
```suggestion
React Native Testing Library allows you to write integration and component tests for your React Native app or library. While the JSX code used in tests closely resembles your React Native app, the things are not quite as simple as they might appear. In this document we will describe the key elements of our testing environment and highlight things to be aware of when writing more advanced tests or diagnosing issues.
```

pierrezimmermannbam commented on website/docs/TestingEnvironment.md:
```diff
@@ -0,0 +1,160 @@
+
+### React renderers
+
+React allows you to write declarative code using JSX, write function or class components, or use hooks like `useState`. In order to output the results of your components it needs to work with a render. Every React app uses some type of renderer: React Native is a renderer for mobile apps, web apps use React DOM, and there are other more [specialised renderers](https://github.com/chentsulin/awesom ...
```
```suggestion
React allows you to write declarative code using JSX, write function or class components, or use hooks like `useState`. In order to output the results of your components it needs to work with a renderer. Every React app uses some type of renderer: React Native is a renderer for mobile apps, web apps use React DOM, and there are other more [specialised renderers](https://github.com/chentsulin/awesome-react-renderer) that can e.g. render to console or HTML canvas.
```

pierrezimmermannbam commented on website/docs/TestingEnvironment.md:
```diff
@@ -0,0 +1,160 @@
+
+React allows you to write declarative code using JSX, write function or class components, or use hooks like `useState`. In order to output the results of your components it needs to work with a render. Every React app uses some type of renderer: React Native is a renderer for mobile apps, web apps use React DOM, and there are other more [specialised renderers](https://github.com/chentsulin/awesom ...
+
+When you run your tests in React Native Testing Library, somewhat contrary to what the name suggest, they are actually **not** using React Native render. This is because this renderer needs to be run on iOS or Android operating system, so it would need to run on device or simulator.
```
```suggestion
When you run your tests in React Native Testing Library, somewhat contrary to what the name suggest, they are actually **not** using React Native renderer. This is because this renderer needs to be run on iOS or Android operating system, so it would need to run on device or simulator.
```

pierrezimmermannbam commented on website/docs/TestingEnvironment.md:
```diff
@@ -0,0 +1,160 @@
+
+### Element tree
+
+Invoking `render()` functions results in creation of element tree. This is done internally by invoking `TestRendere.create()` method. The output tree represents your React Native component tree, each node of that tree is an “instance” of some React component (to be more precise: each node represents a React fiber, and only class components have instances, while function components store the hook  ...
```
```suggestion
Invoking `render()` function results in creation of an element tree. This is done internally by invoking `TestRenderer.create()` method. The output tree represents your React Native component tree, each node of that tree is an “instance” of some React component (to be more precise: each node represents a React fiber, and only class components have instances, while function components store the hook state using fiber).
```

pierrezimmermannbam commented on website/docs/TestingEnvironment.md:
```diff
@@ -0,0 +1,160 @@
+- `*ByDisplayValue` queries returns composite `TextInput` element
+- `*ByPlaceholderText` queries returns composite `TextInput` element
+
+This will change in the near future, as we make efforts for all queries to return host components. Meanwhile it should be a huge issue, as composite `Text` and `TextInput`generally pass their props down to host counterparts.
```
```suggestion
This will change in the near future, as we make efforts for all queries to return host components. Meanwhile it shouldn't be a huge issue, as composite `Text` and `TextInput`generally pass their props down to host counterparts.
```

pierrezimmermannbam commented on website/docs/TestingEnvironment.md:
````diff
@@ -0,0 +1,160 @@
+}
+```
+
+In the above example user defined components accepts both `onPress` and `style` props but does not pass it (through `Pressable`) to host views, so these props will not affect the user interface.
````
I would have added here that the prop names aren't always the same, for instance onClick vs onPress or disabled vs accessibilityState : disabled

pierrezimmermannbam reviewed (changes requested):
Really interesting doc, I think it's a great addition ! I'm just not sure the tree navigation part is relevant, users should be able to use only RNTL api or @testing-library/jest-native matchers to write tests, even though understanding how it works is definitely useful

thymikee commented on website/docs/TestingEnvironment.md:
````diff
@@ -0,0 +1,160 @@
+If you want to find a host element for given element they you might use following code:
+
+```jsx
+function getHostParent(element: ReactTestInstance) {
````
I think we don't ever want this kind of code to end up in the user space, do we?

mdjastrzebski replied:
Removed that code.

thymikee commented on website/docs/TestingEnvironment.md:
```diff
@@ -0,0 +1,160 @@
+title: Testing Environment
+---
+
+## Testing Environment
```
I read through all of that and it's great deep dive into how things work under the hood! However, this is definitely for more advanced users, or when someone needs to debug the problems in this library or their tests. This gives me a hint, that we should be very explicit about it. And either add a note, like React docs does in many places, or make it, idk, a contributing or advanced section :)

thymikee commented on website/docs/TestingEnvironment.md:
````diff
@@ -0,0 +1,160 @@
+
+These tree elements are represented by `ReactTestInstance` type:
+
+```jsx
````
```suggestion
```tsx
```

mdjastrzebski commented:
@thymikee @AugustinLF @pierrezimmermannbam I've moved document to new `Advanced` sidebar section, added a note indicating that this is for more advanced users, and reworked the part on tree navigation.

AugustinLF commented:
@mdjastrzebski did you push your new commit?^^' I don't understand where this `Advanced` sidebar section is.

mdjastrzebski commented:
@AugustinLF LOL, nope ;-)
